### PR TITLE
Baselined to node-red implementation of project

### PR DESCRIPTION
**Pull request by genestealer**
Fully tested with eTRV (MIHO013), Smart Plug Plus (MIHO005) and Smart Monitor Plug (MIHO004).
With a ENER314-RT in SPI mode, on a Raspberry Pi 4. Running NodeJS 16.

## **`app.js`**
1. Added extra text and link to repo.
2. 2. Formatted JS for improved readability.

## **`1.json` MIHO004**
1. Added "Friendly Names" for sensors. 
3. Added missing FREQUENCY and VOLTAGE. 
4. Rounded FREQUENCY
image

## **`2.json` MIHO005**
1. Added "Friendly Names" for everything. 
2. Added missing FREQUENCY and VOLTAGE. 
3. Rounded FREQUENCY
5. Added device class.
6. Added payload and state values.
7. Added diagnostic binary sensor to show the returned switch state.
image

## **`3.json` MIHO013**
1. Added "Friendly Names" for everything. 
2. Added temp_step 0.5 for 1/2 degree temp steps
3. Formatted for improved readability.
4. Removed "icon":"mdi:battery" to enable device class to allow home assistant to show dynamic icon.
5. Moved Maintenance and REPORTING_INTERVAL to Configuration category.
image

### CODE REVIEW

File: README.md
Comment (Achronite) [inline README.md:1]: Are these changes the same as the PR I just merged?

File: devices/1.json
Comment (Achronite) [inline devices/1.json:1]: Do you have a need for Frequency & Voltage to be recorded?  I found they just got in the way as they never change.

File: README.md
Comment (genestealer) [reply README.md:1]: No, the PR you just merged (https://github.com/Achronite/mqtt-energenie-ener314rt/pull/45) was just for the readme, this is for the updates mostly to the front-end via MQTT discovery.

File: devices/1.json
Comment (genestealer) [reply devices/1.json:1]: Are you sure, that's surprising, in Worcestershire they can vary significantly, I have automated alerts if they go out of spec and call up the national grid, as we have seen the local transformer variable tap get stuck and the voltage peek and sag way out of acceptable ranges. See below:

image
image


If you would rather not show then, we can add `"enabled_by_default":"false"}` to the sensor. So they will show in Home Assistant, but will have to be manually enabled by the user in Home Assistant.

```
{"id": "FREQUENCY",     "component":"sensor", "stat_t":"~@/state","config": {"name":"Frequency","value_template":"{{ value|round(2) }}","dev_cla":"frequency","unit_of_meas":"Hz","enabled_by_default":"false"}},
{"id": "VOLTAGE",     "component":"sensor", "stat_t":"~@/state","config": {"name":"Voltage","dev_cla":"voltage","unit_of_meas":"V","enabled_by_default":"false"}}
```
 
 
 Looks like this:
image
image
image
image

File: package.json
Comment (Achronite) [inline package.json:1]: I've not decided what version this will be yet.  It wont be v0.5.1

Comment (genestealer) [reply package.json:1]: No worries, I thought it was presumptuous for me to make the change myself so I was putting it back to leave up to you

**Comment by Achronite** [comment]
Just realised these changes are against the master branch.  I stage all development on 'develop' first before merging into master upon packaged release.  Happy for you to change it over, or I'll do it later.

**Comment by genestealer** [comment]
> Just realised these changes are against the master branch.  I stage all development on 'develop' first before merging into master upon packaged release.  Happy for you to change it over, or I'll do it later.

Only took my second to change it over. Done.

File: devices/1.json [hunk 1]
```diff
@@ -1,6 +1,8 @@
 {"mdl":"Monitor Plug",
     "parameters":[
-        {"id": "REAL_POWER",    "component":"sensor", "stat_t":"~@/state","config": {"dev_cla":"power","unit_of_meas":"W"}},
-        {"id": "REACTIVE_POWER","component":"sensor", "stat_t":"~@/state","config": {"dev_cla":"reactive_power","unit_of_meas":"var"}}
+        {"id": "REAL_POWER",    "component":"sensor", "stat_t":"~@/state","config": {"name":"Power","dev_cla":"power","unit_of_meas":"W","enabled_by_default":"true"}},
+        {"id": "REACTIVE_POWER","component":"sensor", "stat_t":"~@/state","config": {"name":"Reactive Power","dev_cla":"reactive_power","unit_of_meas":"var","enabled_by_default":"true"}},
+        {"id": "FREQUENCY",     "component":"sensor", "stat_t":"~@/state","config": {"name":"Frequency","value_template":"{{ value|round(2) }}","dev_cla":"frequency","unit_of_meas":"Hz","enabled_by_default":"true"}},
+        {"id": "VOLTAGE",     "component":"sensor", "stat_t":"~@/state","config": {"name":"Voltage","dev_cla":"voltage","unit_of_meas":"V","enabled_by_default":"true"}}
```
Comment (Achronite) [inline devices/1.json:6]: `enabled_by_default` should only be specified when not default (true). It should be specified as `false` for Voltage and Frequency for backwards compatibility.
Also, I'm not sure about `name` being specified here; it ends up being defined twice for the entities in the discovery message. `app.js` already adds it (albeit lowercase at present - but I have a fix for that :) )

File: devices/2.json [hunk 2]
```diff
@@ -1,6 +1,20 @@
...
     "parameters":[
-        {"id": "switch",        "component":"switch", "stat_t":"~@/state","cmd_t":"~@/command", "config": {"icon":"mdi:power-socket-uk","opt":false}},
-        {"id": "REAL_POWER",    "component":"sensor", "stat_t":"~@/state","config": {"dev_cla":"power","unit_of_meas":"W"}},
-        {"id": "REACTIVE_POWER","component":"sensor", "stat_t":"~@/state","config": {"dev_cla":"reactive_power","unit_of_meas":"var"}}    ]
+        {"id": "switch", "component":"switch", "stat_t":"~@/state", "cmd_t":"~@/command", "config": {
+            "name":"Switch",
+            "dev_cla":"outlet",
+            "icon":"mdi:power-socket-uk",
+            "opt":false,
+            "payload_on":"ON",
```
Comment (Achronite) [inline devices/2.json:8]: payload_on, payload_off, state_on and state_off are all default values and shouldn't be specified here

Comment (Achronite) [reply devices/2.json:8]: Also see comments for 1.json and apply here

Comment (Achronite) [reply devices/2.json:8]: remove id "Switch" it is already defined

File: devices/2.json [hunk 3]
```diff
@@ -1,6 +1,20 @@
...
+            "payload_off":"OFF",
+            "state_on":"ON",
+            "state_off":"OFF",
+            "enabled_by_default":"true"
+            }},
+        {"id": "REAL_POWER",    "component":"sensor", "stat_t":"~@/state","config": {"name":"Power","dev_cla":"power","unit_of_meas":"W","enabled_by_default":"true"}},
+        {"id": "REACTIVE_POWER","component":"sensor", "stat_t":"~@/state","config": {"name":"Reactive Power","dev_cla":"reactive_power","unit_of_meas":"var","enabled_by_default":"true"}},
+        {"id": "FREQUENCY",     "component":"sensor", "stat_t":"~@/state","config": {"name":"Frequency","value_template":"{{ value|round(2) }}","dev_cla":"frequency","unit_of_meas":"Hz","enabled_by_default":"true"}},
+        {"id": "VOLTAGE",       "component":"sensor", "stat_t":"~@/state","config": {"name":"Voltage","dev_cla":"voltage","unit_of_meas":"V","enabled_by_default":"true"}},
+        {"id": "switch",        "component":"binary_sensor", "stat_t":"~@/state","config": {"name":"Switch State","dev_cla":"power","entity_category":"diagnostic", "payload_on":"ON", "payload_off":"OFF","enabled_by_default":"true"}}
```
Comment (Achronite) [inline devices/2.json:18]: remove id "Switch" it is already defined

File: app.js [hunk 4]
```diff
@@ -724,10 +724,27 @@ function publishDiscovery( device, index ){
...
-											uniq_id: `${unique_id}`, "~": `${CONFIG.topic_stub}${device.productId}/${device.deviceId}/`, name: `${entity_name}`, availability_topic: `${CONFIG.topic_stub}availability/state` },
-											parameter.config,);
+
+					var dmsg = Object.assign({
+						device: {
+							name: `${device_name}`,
+							ids: [`ener314rt-${device.deviceId}`],
+							mdl: `${device_defaults.mdl}`,
+							mf: `Energenie`,
+							sw: `${APP_VERSION}`,
```
Comment (Achronite) [inline app.js:734]: I've already made the change to add APP_VERSION under sw

Comment (Achronite) [reply app.js:734]: + remove `hw` line

File: app.js [hunk 5]
```diff
@@ -724,10 +724,27 @@ function publishDiscovery( device, index ){
...
+							hw: `mqtt-ener314rt`
+						},
+						uniq_id: `${unique_id}`,
+						"~": `${CONFIG.topic_stub}${device.productId}/${device.deviceId}/`,
+						name: `${entity_name}`,
+						availability_topic: `${CONFIG.topic_stub}availability/state`,
+						origin: {
```
Comment (Achronite) [inline app.js:741]: Not seen `origin` before.  Where is the documentation for this?

Comment (genestealer) [reply app.js:741]: https://www.home-assistant.io/integrations/mqtt/
image
image

Comment (Achronite) [reply app.js:741]: OK, pls use the abbreviations though, `o` etc, it cuts down on network traffic and MQTT usage.

File: app.js [hunk 6]
```diff
@@ -724,10 +724,27 @@ function publishDiscovery( device, index ){
...
+							name: `mqtt-ener314rt with ENER314-RT Raspberry Pi Shield`,
```
Comment (Achronite) [inline app.js:742]: Please use `mqtt-energenie-ener314rt` as name

File: devices/1.json (hunk 1 above)
Comment (Achronite) [reply devices/1.json:6]: I've pushed the change to develop for the entity names to be auto-constructed Title Case (from lower case), this should remove your need for the "friendly names" in the .json files.

File: app.js (hunk 6 above)
Comment (genestealer) [reply app.js:742]: Done

File: app.js (hunk 5 above)
Comment (genestealer) [reply app.js:741]: Done

File: app.js (hunk 4 above)
Comment (genestealer) [reply app.js:734]: > * remove `hw` line

OK

File: devices/1.json (hunk 1 above)
Comment (genestealer) [reply devices/1.json:6]: > I've pushed the change to develop for the entity names to be auto-constructed Title Case (from lower case), this should remove your need for the "friendly names" in the .json files.

There was a but with your code, but I have fixed it to also auto-constructed Title Case from the lower case words.

File: devices/2.json (hunk 2 above)
Comment (genestealer) [reply devices/2.json:8]: done

File: devices/2.json (hunk 3 above)
Comment (genestealer) [reply devices/2.json]: @Achronite regarding the (now removed) binary_sensor switch, how would you create such as switch if you needed it, if not via the means I did?

File: app.js [hunk 7]
```diff
@@ -724,10 +724,25 @@ function publishDiscovery( device, index ){
...
-					var dmsg = Object.assign( { device: { name: `${device_name}`, ids: [`ener314rt-${device.deviceId}`], mdl: `${device_defaults.mdl}`, mf: 'Energenie', sw: `mqtt-ener314rt ${APP_VERSION}` }, 
-											uniq_id: `${unique_id}`, "~": `${CONFIG.topic_stub}${device.productId}/${device.deviceId}/`, name: `${entity_name}`, availability_topic: `${CONFIG.topic_stub}availability/state` },
-											parameter.config,);
+					var dmsg = Object.assign({
+						device: {
+							name: `${device_name}`,
+							ids: [`ener314rt-${device.deviceId}`],
+							mdl: `${device_defaults.mdl}`,
+							mf: `Energenie`,
+							sw: `${APP_VERSION}`
```
Comment (Achronite) [inline app.js:733]: Can you leave sw as is - sw: `mqtt-ener314rt ${APP_VERSION}`

Comment (genestealer) [reply app.js:733]: > Can you leave sw as is - sw: `mqtt-ener314rt ${APP_VERSION}`

Reverted to `- sw: mqtt-ener314rt ${APP_VERSION}`

**Comment by genestealer** [comment]
Will open new pull request for review

**Pull request closed without merging**